R/convergence.R

```diff
@@ -316,6 +316,144 @@ quantile2 <- function(x, probs = c(0.05, 0.95), names = TRUE, ...) {
+#'
+#' @param split_chains a Boolean indicating whether to split chains into two equal
+#' halves.
+#'
+#' @param uncertainty whether to provide a vector of R* values (if true) or a single
+#' value (if false).
+#'
+#' @param method machine learning classifer available in caret R package.
+#'
+#' @param hyperparameters hyperparameter settings for classifier given as a list.
+#'
+#' @param nsim number of R* values in returned vector if uncertainty is true.
+#'
+#' @param training_proportion proportion of iterations used to train classifier.
+#'
+#' @details `rstar` provides a measure of MCMC convergence based on whether it is
+#' possible to determine the chain that generated a particular draw with a probability
+#' greater than chance. To do so, it fits a machine learning classifier to a subset of
+#' the original MCMC draws (the training set) and evaluates its predictive accuracy on
+#' the remaining draws (the testing set). If predictive accuracy exceeds chance (i.e.
+#' predicting the chain that generated a draw uniformly at random), the diagnostic
+#' measure, R* > 1, indicating that convergence has yet to occur.
+#'
+#' The statistic, R*, is stochastic, meaning that each time the test is run, unless
+#' the random seed is fixed, it will generally produce a different result. To minimise
+#' the implications of this stochasticity, it is recommended to repeatedly run this
+#' function to calculate a distribution of R*; alternatively, an approximation to this
+#' distribution can be obtained by setting uncertainty = T.
+#'
+#' By default, a random forest classifier is used, which tends to perform best for
+#' target distribution dimensionalities of around 4 and above. For lower dimensional
+#' targets, gradient boosted models (called via `method=\"gbm\"`) tend to have a
+#' higher classification accuracy. On a given MCMC sample, it is recommended to
+#' try both of these classifiers.
+#'
+#' @return A single numeric R* value (by default) or a vector of values (if
+#' uncertainty is true).
+#'
+#' @references Ben Lambert, Aki Vehtari (2020) R*: A robust MCMC convergence
+#' diagnostic with uncertainty using gradient-boosted machines
+#' \emph{arXiv preprint} \code{arXiv:TBD}
+#'
+#' @examples
+#' x <- example_draws("eight_schools")
+#' rstar(x)
+#' rstar(x, split_chains = F)
+#' rstar(x, method = "gbm")
+#' # can pass additional arguments to methods
+#' rstar(x, method = "gbm", verbose = F)
+#'
+#' # with uncertainty, returns a vector of R* values
+#' hist(rstar(x, uncertainty = T))
+#' hist(rstar(x, uncertainty = T, nsim = 100))
+#'
+#' # can use other classification methods in caret library
+#' rstar(x, method = "knn")
+#'
+#' @export
+rstar <- function(x, split_chains=T, uncertainty=F, method=NULL, hyperparameters=NULL,
+                            training_proportion=0.7, nsim=1000, ...){
+
+  if(!requireNamespace("caret", quietly = T))
+    stop("Package \"caret\" needed for this function to work. Please install it.")
+  require(caret)
```

Comment:
Would it be possible to access caret functions via `caret::` rather than calling `require(caret)`?

Reply:
caret has some annoying features; one of them seems to be that, to access particular ML methods, the package itself needs to be loaded. For example, when I do:

`fit <- caret::train(.chain ~ .,
                      data = training_data,
                      method = "knn",
                      trControl = caret::trainControl(method = 'none'),
                      tuneGrid = caret_grid,
                      ...)`

I get an error message if caret isn't loaded. Whereas:

`
require(caret)
fit <- train(.chain ~ .,
                      data = training_data,
                      method = "knn",
                      trControl = trainControl(method = 'none'),
                      tuneGrid = caret_grid,
                      ...)`

works. What do you think?

Reply:
I see. I have no experience with caret myself and I would think that `require(caret)` is ok in this case if we cannot use `::` reliably.

Reply:
Thanks. Ok -- will keep as is.

Reply:
I might be wrong but I would think using `require()` inside package code would be discouraged by CRAN. I know that using `library()` is either not allowed or strongly discouraged, so I would think they would similarly discourage using `require()`. The reason, I assume, is that calling a function in one package shouldn't result in another package being loaded because that affects the global environment. That is, once we call `require(caret)` then caret is loaded for the user even after the `rstar()` function is exited, which can result in name clashes and other unexpected behavior.

Reply:
Yeah. You are right. This is what CRAN says (https://cran.r-project.org/doc/manuals/r-release/R-exts.html):

"R code in the package should call library or require only exceptionally. Such calls are never needed for packages listed in ‘Depends’ as they will already be on the search path. It used to be common practice to use require calls for packages listed in ‘Suggests’ in functions which used their functionality, but nowadays it is better to access such functionality via :: calls. "

We could 

1) state in the comments to CRAN that the `require` call is necessary to make caret work. 
2) figure out a workaround which works with `::` calls (not sure how though). 
3) file an issue for caret to fix its use when not loaded and hope it will be fixed soon.

Reply:
What if we check if caret is loaded (on the search path?) and if not ask the user to manually load it? Would that work?

Reply:
@jgabry I like your idea. Will try to do this later tonight.

Reply:
Thanks!

Reply:
I've changed this now but it's failing because Github actions isn't finding the caret, gbm, randomForest and e1071 packages. These are suggests which previously I was requiring in Travis. Any idea how to handle these in Github actions?

Reply:
Not 100% sure but I think the place to do it is probably here: 

https://github.com/stan-dev/posterior/blob/cfa6dc5f04195d6b450270818ccb04a858f7bfec/.github/workflows/rcmdcheck.yml#L74-L77

The `remotes::install_cran()` function can take a vector of package names.